### PR TITLE
prober: make DERP probes fail when stun requests fail

### PR DESCRIPTION
**Pull request by raggi**
The client treats a single lost stun request as a failure in netcheck and will bounce regions as a result. The prober results should represent this if it is happening regularly in a region we need to know about that. The fist step is to report the data with more precision; the likely immediate next step is to adjust alerting to soften the blow.

Updates tailscale/corp#11492

### CODE REVIEW

File: prober/derp.go [hunk 1]
```diff
@@ -209,47 +210,44 @@ func derpProbeUDP(ctx context.Context, ipStr string, port int) (latency time.Dur
...
-		}
-		if txBack != tx {
-			return 0, fmt.Errorf("read wrong tx back from %v", ip)
-		}
-		if latency == 0 || d < latency {
-			latency = d
+	addr := netip.AddrPortFrom(ip, uint16(port))
+
+	// Binding requests and responses are fairly small (~40 bytes),
+	// but in practice a STUN response can be up to the size of the
```
Comment (danderson) [inline prober/derp.go:223]: Why can it be that? At least from our stun servers the size should be static and small, shouldn't it?

**Review by danderson — APPROVED** [review]

File: prober/derp.go [hunk 2]
```diff
@@ -209,47 +210,44 @@ func derpProbeUDP(ctx context.Context, ipStr string, port int) (latency time.Dur
...
+	_, err = uc.WriteToUDPAddrPort(req, addr)
+	if err != nil {
+		return 0, err
+	}
+	t0 := time.Now()
+	n, _, err := uc.ReadFromUDP(buf)
+	d := time.Since(t0)
+	if err != nil {
+		if ctx.Err() != nil || err == context.DeadlineExceeded || d > time.Second {
+			return 0, fmt.Errorf("timeout reading from %s: (%s) %w", addr, d, err)
```
Comment (valscale) [inline prober/derp.go:238]: Tiny nit: it seems like "d > time.Second" could mask a different error, the one recorded in err.

File: prober/derp.go [hunk 3]
```diff
@@ -209,47 +210,44 @@ func derpProbeUDP(ctx context.Context, ipStr string, port int) (latency time.Dur
...
+		return 0, fmt.Errorf("error reading from %s: (%s) %w", addr, d, err)
+	}
+	txBack, _, err := stun.ParseResponse(buf[:n])
+	if err != nil {
+		return 0, fmt.Errorf("parsing STUN response from %s: %v", addr, err)
+	}
+	if txBack != tx {
+		return 0, fmt.Errorf("read wrong tx back from %s", addr)
+	}
+	if latency == 0 || d < latency {
```
Comment (valscale) [inline prober/derp.go:249]: I think this can now be just 'latency = d'?

File: prober/derp.go [hunk 4]
```diff
@@ -209,47 +210,44 @@ func derpProbeUDP(ctx context.Context, ipStr string, port int) (latency time.Dur
...
-		}
-		if latency == 0 || d < latency {
-			latency = d
+	addr := netip.AddrPortFrom(ip, uint16(port))
+
+	// Binding requests and responses are fairly small (~40 bytes),
+	// but in practice a STUN response can be up to the size of the
+	// path MTU, so we use a jumbo frame size buffer here.
+	buf := make([]byte, 9000)
+
```
Comment (valscale) [inline prober/derp.go:226]: I notice this removes the 'uc.SetReadDeadline(time.Now().Add(2 * time.Second))' from the previous code - just checking if there is a default deadline/timeout?

**Review by valscale — COMMENTED** [review]
Thanks for fixing this!

**Pull request closed without merging**